packages/jupyter-ai/src/chat_handler.ts

```diff
@@ -0,0 +1,167 @@
+import { IDisposable } from '@lumino/disposable';
+import { ServerConnection } from '@jupyterlab/services';
+import { URLExt } from '@jupyterlab/coreutils';
+import {Poll} from '@lumino/polling';
+
+
+const API_NAMESPACE = 'api/ai';
+
+/**
+ * Call the API extension
+ *
+ * @param endPoint API REST end point for the extension
+ * @param init Initial values for the request
+ * @returns The response body interpreted as JSON
+ */
+export async function requestAPI<T>(
+  endPoint = '',
+  init: RequestInit = {}
+): Promise<T> {
+  // Make request to Jupyter API
+  const settings = ServerConnection.makeSettings();
+  const requestUrl = URLExt.join(settings.baseUrl, API_NAMESPACE, endPoint);
+
+  let response: Response;
+  try {
+    response = await ServerConnection.makeRequest(requestUrl, init, settings);
+  } catch (error) {
+    throw new ServerConnection.NetworkError(error as TypeError);
+  }
+
+  let data: any = await response.text();
+
+  if (data.length > 0) {
+    try {
+      data = JSON.parse(data);
+    } catch (error) {
+      console.log('Not a JSON response body.', response);
+    }
+  }
+
+  if (!response.ok) {
+    throw new ServerConnection.ResponseError(response, data.message || data);
+  }
+
+  return data;
+}
+
+const CHAT_SERVICE_URL = "api/ai/chats"
+
+export class ChatHandler implements IDisposable{
+  /**
+   * Create a new event manager.
+   */
+  constructor(options: ChatHandler.IOptions = {}) {
+    this.serverSettings =
+      options.serverSettings ?? ServerConnection.makeSettings();
+
+    this._poll = new Poll({ factory: () => this._subscribe() });
+    this._poll.start();
+  }
+
+  /**
+   * The server settings used to make API requests.
+   */
+  readonly serverSettings: ServerConnection.ISettings;
+
+  /**
+   * Whether the chat handler is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+   /**
+   * Dispose the chat handler.
+   */
+   dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this._isDisposed = true;
+
+    // Clean up poll.
+    this._poll.dispose();
+
+    this._listeners = []
+
+    // Clean up socket.
+    const socket = this._socket;
+    if (socket) {
+      this._socket = null;
+      socket.onopen = () => undefined;
+      socket.onerror = () => undefined;
+      socket.onmessage = () => undefined;
+      socket.onclose = () => undefined;
+      socket.close();
```

Comment:
Do we need to unbind the hooks prior to calling `socket.close()`? Seems redundant.

Reply:
I derived this from EventsManager which had this cleanup. I am assuming this takes care of phantom sockets in certain browsers, even when they are closed.